### ISSUE TITLE
pnext: API for accessing handle table generation

### ISSUE THREAD

**Issue by eriohl** [Starcounter/Starcounter#3648]
Per context Generation and Generation index last updated + First generation references all contexts (entire system).

cc @k-rus @alexanderstc

**Comment by eriohl** [comment]
Function ```uint32_t stari_context_get_snapshot_info(uint64_t handle, uint64_t *pgeneration, uint64_t *pindex_generation);``` added. Get information about the snapshot used by the current transaction.

**Comment by eriohl** [comment]
I'm rewriting the how this with determining first referenced generation works so unless there is an urgent need to get the function for determining first reference generation I'm planning to postpone implementing this until this rewrite is in place?

Also is if system wide first referenced generation or IWP first referenced generation you want? I stated the former in the task description but thinking about it I guess what you actually need is the first referenced for the IWP right? If seems to me that it doesn't matter if another IWP references a ealier generation since it will manage its own cache.

**Comment by alexanderstc** [comment]
@k-rus how urgent is this?
I'd guess it's not very urgent.

Currently caching is per scdata's "client". What exactly IWP stands for, to be sure?

I was thinking about eventually investigating sharing cache between "clients" to save on compilation, but that's purely hare-brain speculation at the moment.

**Comment by eriohl** [comment]
IWP = Independent Worker Process = scdata client.

> I was thinking about eventually investigating sharing cache between "clients" to save on compilation, but that's purely hare-brain speculation at the moment.

I'll do one function that gets IWP and one that gets system wide and then you can choose which one you want. Minimal amount of extra work.

**Comment by alexanderstc** [comment]
Thanks!

**Comment by k-rus** [comment]
It is not urgent. The work on cache will start after the work on CRUD API.
Since creating/dropping index is system wide operation, I guess the first cache implementation will be wide system too. In long term of AppIOS the situation can be changed.

**Comment by alexanderstc** [comment]
I'm starting to wait on this.

**Comment by eriohl** [comment]
I've added interface function ```stari_get_local_first_referenced_generation``` that gets the first referenced generation of the current process.

As discussed offline: Getting the system wide first referenced generation is not kind of hard to do because there is currently now way for one IWP to see first referenced generation of another IWP. So if we need that we need to rework things a bit (quite a bit actually). But unless the idea is to share the query cache between multiple IWPs, which I gather it is not, there should be no need for this for the purpose of cleaning up unused query plans. Right?

Anyway: I'm going to spend a bit of time on cleaning this up and testing it tomorrow (and it might be renamed). But I expect it will work.

**Comment by eriohl** [comment]
When writing this I started to wonder if it is the index generation and not the generation we need for this? Index generation can be (and mostly is) lower the the generation.

**Comment by alexanderstc** [comment]
Terminology: "lower" means ealier/older, right?

**Comment by eriohl** [comment]
Yes

**Comment by alexanderstc** [comment]
I hope using your api to know what is the earliest/oldest generation in play. Using that I hope to track what compiled queries cache's entries I can erase.
If your api can provide separate info on tables and indexes, I'll use it to my advantage to make the tracking more precise (not sure how much difference it will make in real life scenarios).
Otherwise I think it would be better to report the oldest generation.

**Comment by eriohl** [comment]
First generation is the oldest generation. Index generation is the låst generation where indexing has changed (which then is more then likely to be earlier then the first referenced generation).

**Comment by alexanderstc** [comment]
> låst generation where indexing has changed

I don't see how I could use that in place of the oldest referenced.
Given this choice I'd go with the oldest referenced generation.

**Comment by eriohl** [comment]
Ok then.

```stari_get_local_first_referenced_generation``` is not so great performance wise so we might need to revisit this depending on how it is used.

Also discovered a bug in index updated callback where the context index was set to the snapshot (handle table) index and not the context index. Fixed now.

**Comment by alexanderstc** [comment]
>  stari_get_local_first_referenced_generation  is not so great performance wise

Noted.

**Comment by alexanderstc** [comment]
I started to test the api and I get the impression:
if I forget to properly release a transaction it hangs around until db restart.
Is that so?

**Comment by eriohl** [comment]
The system is unable to GC a generation referenced by an snapshot isolation transaction. This lock on GC will be released however if the IWP that is holding the transaction open is terminated however (which is not exactly a db restart). Is that what you were asking?

**Comment by alexanderstc** [comment]
> Is that what you were asking?

Yes, thank you.

I don't mean to criticize or anything, but isn't this relying on client's good behaviour?
(Mine metalayer_functest seems to misbehave a lot, I'll fix it, of course).

**Comment by eriohl** [comment]
I agree that it makes the system a bit more fragile (although the system has always relied on the workers to behave to some degree, now less then before). Discussed it a bit [here](https://github.com/Starcounter/CompanyTrack/issues/196). But I reasoned that, because of the compatibility break, it was more important to get the snapshot working with the current transaction scheduling quickly rather then solving every possible issue arising from that first. In the longer term we will have to make improvements to GC to compensate and possibly also introduce some transaction timeout or something. For now I'm told that it works as it should.

**Comment by alexanderstc** [comment]
> added interface function  stari_get_local_first_referenced_generation  that gets the first referenced generation of the current process

Just to be absolutely clear: is that the earliest referenced of both index and "non-index" generations?

**Comment by eriohl** [comment]
> > added interface function stari_get_local_first_referenced_generation that gets the first referenced generation of the current process

> Just to be absolutely clear: is that the earliest referenced of both index and "non-index" generations?

As I commented previously: First generation is the oldest generation. Index generation is the last generation where indexing has changed (which then is more then likely to be earlier then the first referenced generation).

For example: Lets say you have this series of transactions:

G1: Update record.
G2: Create index.
G3: Update record.

Then if all transactions would reference the latest generation (3) then this function would output 3. Index generation would however be 2.

Hence my comment that it might be the index generation you want in this case.

**Comment by alexanderstc** [comment]
I think I need both, could you upgrade stari_get_local_first_referenced_generation() to return the index generation as well?

**Comment by eriohl** [comment]
Kind of tricky since it isn't tracked the same way. But I'll look into it.

**Comment by alexanderstc** [comment]
I'd like to remind that I only need a way to know if an object (index or layout), deleted in a particular transaction, is no longer referenced by other transactions.

**Comment by eriohl** [comment]
> I'd like to remind that I only need a way to know if an object (index or layout), deleted in a particular transaction, is no longer referenced by other transactions.

From this comment it, for me, became kind of less clear what you actually want.

But for indexes you can determine when (in which generation) an index is dropped (although we would have to add an API function for this) and you can determine if this generation is still referenced with this API. Would this be sufficient?

For layouts I don't really understand what you mean since layouts are only deleted when it can be determined that there are no references to it (including transactions)?

**Comment by alexanderstc** [comment]
> you can determine when (in which generation) an index is dropped (although we would have to add an API function for this) 

I plan track that myself in metalayer index delete function, can leave without api function. However, if it's really, really easy to add that function, that would be nice too.

> an index is dropped ...  and you can determine if this generation is still referenced with this API. Would this be sufficient?

Yes, thank you!

> For layouts I don't really understand what you mean since layouts are only deleted when it can be determined that there are no references to it (including transactions)?

Again, I track not the actual layout deletion, but metalayer-level table deletion in the metalayer drop_table function, I already have everything I need there, no additional support is required for tables.

My comment about tracking kernel object is therefore incorrect regarding layouts.

**Comment by eriohl** [comment]
> > an index is dropped ... and you can determine if this generation is still referenced with this API. Would this be sufficient?

> Yes, thank you!

New function ```stari_context_get_index_generation_dropped``` added. Sorry it took a while.
